Nested Map Functions

HamptonMakes opened the pull request:
Now, map-merge, map-get, and map-has-key all accept various forms of nested Map logic as per @nex3's specifications in #1739, with the addition of a nested compatible map-has-key.

saper committed:
Documentation: Fix anchor to Sass options

Sass options live now under

    http://sass-lang.com/documentation/file.SASS_REFERENCE.html#options
and not

    http://sass-lang.com/documentation/file.SASS_REFERENCE.html#sass_options

HamptonMakes committed:
Implemented nested map functions #1739

Now, map-merge, map-get, and map-has-key all accept various forms of nested Map logic as per @nex3's specifications in #1739, with the addition of a nested compatible map-has-key.

HamptonMakes committed:
update declarations as var-args, and fix name of map remove, and add examples of multi-key map-has-key usage

HamptonMakes committed:
add notes about the new nested map methods into the changelog

nex3 commented on lib/sass/script/functions.rb:
```diff
@@ -196,11 +196,13 @@ module Sass::Script
   #
-  # \{#map_get map-get($map, $key)}
-  # : Returns the value in a map associated with a given key.
+  # \{#map_get map-get($map, $keys)}
```
`$keys...` here and elsewhere.

nex3 commented on lib/sass/script/functions.rb:
```diff
@@ -2053,15 +2055,23 @@ def list_separator(list)
+      result || null
     end
-    declare :map_get, [:map, :key]
+    declare :map_get, [:map, :key], :var_args => true
```
Get rid of the `:key` declaration here—the the variable argument name can't be used as a named parameter anyway (see `call()` for another example of the same thing).

You should also preserve the previous declaration, so that `map-get()` calls with named parameters continue to work.

nex3 commented on lib/sass/script/functions.rb:
```diff
@@ -2078,17 +2088,35 @@ def map_get(map, key)
+    #   @param $value [Sass::Script::Value::Base]
+    # @return [Sass::Script::Value::Map]
+    # @raise [ArgumentError] if the first parameter is not a map
+    def map_merge(map1, *args)
```
[I've changed my mind on how this should work.](https://github.com/sass/sass/issues/1739#issuecomment-156571180)

nex3 commented on lib/sass/script/functions.rb:
```diff
@@ -2053,15 +2055,23 @@ def list_separator(list)
-      map.to_h[key] || null
+      result = map
+      keys.each do |key|
+        if result.is_a?(Sass::Script::Value::Map) || result.is_a?(Hash)
```
Why would `result` ever be a Ruby hash here?

nex3 commented on lib/sass/script/functions.rb:
```diff
@@ -2147,16 +2175,19 @@ def map_values(map)
       assert_type map, :Map, :map
-      bool(map.to_h.has_key?(key))
+      bool(keys.all? do |key|
+        map = map.to_h[key] || false
```
`nil` is falsey, so you don't need the `|| false` here.

Also, what if `map` is no longer a map?

nex3 commented on lib/sass/script/value/map.rb:
```diff
@@ -43,6 +43,17 @@ def eq(other)
       Bool.new(other.is_a?(Map) && value == other.value)
     end
 
+    def surgical_merge(my_key, *keys)
```
Please document any methods you add.

nex3 commented on test/sass/functions_test.rb:
```diff
@@ -1357,10 +1357,20 @@ def test_unique_id
+
+    # map-get should return a 'null' if the nested map key isn't found
+    assert_equal "null", perform("map-get((a: b), a, b, c, m)").to_sass
+    assert_equal "null", perform("map-get((a: (b: (c: d))), a, b, c, m)").to_sass
```
Test accessing a nested empty map.

nex3 commented on test/sass/functions_test.rb:
```diff
@@ -1419,8 +1439,10 @@ def test_map_values_checks_type
+    assert_equal "true", evaluate("map-has-key((foo: (bar: 1)), foo, bar)")
     assert_equal "false", evaluate("map-has-key((foo: 1, bar: 1), baz)")
     assert_equal "false", evaluate("map-has-key((), foo)")
+    assert_equal "false", evaluate("map-has-key((foo: (bar: 1)), foo, bing)")
```
Test nesting deeper than actually exists, nesting through values that aren't maps, and nesting through empty maps.

HamptonMakes committed:
Integrating in the changes requested in @nex3\'s comments

Plus using the new specification given for map-merge, including the creation of a new `map-set` function
